<!doctype html>
<html>
	<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="chrome=1">
		<title>Genome browser NG by ohtubrowser</title>

		<link rel="stylesheet" href="stylesheets/styles.css">
		<link rel="stylesheet" href="stylesheets/pygment_trac.css">
		<script src="javascripts/scale.fix.js"></script>
		<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

		<!--[if lt IE 9]>
		<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
		<![endif]-->
	</head>
	<body>
		<div class="wrapper">
			<header class="without-description">
			<h1>Genome Browser NG</h1>
			<p></p>
			<p class="view"><a href="https://github.com/ohtubrowser/gbrowserng">View the Project on GitHub <small>ohtubrowser/gbrowserng</small></a></p>
			<!--
			<ul>
				<li><a href="#index"><strong>Index</strong></a></li>
				<li><a href="#howto"><strong>How to use</strong></a></li>
				<li><a href="#download"><strong>Download</strong></a></li>
			</ul>
			-->
			</header>
			<section>
			<nav id="index"><a href="#index">Index</a> - <a href="#video">Usage video</a> - <a href="#howto">How to use</a> - <a href="#download">Download</a></nav>
			<h2>What is Genome Browser NG?</h2>
			<div id="mainscr"><a href="images/mainscr.png"><img src="images/mainscr.png" height="400px" alt="Screenshot"/></a></div>
			<p>This project develops genome browser software that builds on game-like user interface design and powerful Java-based OpenGL graphics libraries. The genome browser uses some of the basic principles found from previous similar browser software, but introduces a new kind of user interface that allows significantly more flexible data navigation.</p>
			<p>
			The high level goal of the project is to develop the genome browser into a such state that it can be used in real life use cases to visualize data on fusion genes. In other words, the target set of functionalities are a focused subset of the general genome browser, but nevertheless, constitute a useful application that is an improvement over the existing genome browser software. The work will be centered on developing further the circular view mode of the prototype.
			</p>
			<p>
			The focus of this project are fusion genes and other chromosomal rearrangements. Chromosomal rearrangements are drastic mutations where large parts of the chromosome have "switched places". The discovery and visualizations of these are of paramount importance to many disciplines in medical research, a prominent example being cancer genetics. Fusion genes are a result of small scale chromosomal rearrangements and have been observed in many cancer types.  Databases like the Catalogue Of Somatic Mutations In Cancer - (COSMIC), currently contain information of more than 7000 such genes. One can distinguish several types of rearrangement events depending on if they are  affecting a single gene or rather a part of, or a whole, chromosome and whether the pieces of DNA have been inverted upon fusion.
			</p>
			<p>
			Genome browsing is based on the idea of a reference genome: to navigate genomic coordinates The reference genome is used to infer coordinates for everything that is shown in the visualization. If the actual genome under study has major rearrangements, it will result in some pairs of sequencing reads to align to completely different locations of the reference genome. For designers of genome browser software, this is a challenge. Almost all data that genome browsers show is strongly local: if you are interested in basepairs 100000-100100 of chromosome 1, then you can view only that area and will see everything relevant. But rearrangements are an exception, because you can have paired reads where one pair is in your view and the other in some completely different, remote, location. Typical browsers have troubles in communicating this to the user.
			</p>
			<nav id="video"><a href="#index">Index</a> - <a href="#video">Usage video</a> - <a href="#howto">How to use</a> - <a href="#download">Download</a></nav>
			<h2>Check out the video!</h2>
			<p>Video coming later!</p>
			<!--
				<p>This is a video from a development version of the program.</p>
				<p><iframe src="http://player.vimeo.com/video/41079437?title=0&amp;byline=0&amp;portrait=0" width="700" height="500" frameborder="0" webkitAllowFullScreen mozallowfullscreen allowFullScreen></iframe></p>
			-->

			<nav id="howto"><a href="#index">Index</a> - <a href="#video">Usage video</a> - <a href="#howto">How to use</a> - <a href="#download">Download</a></nav>
			<h2>How do I use this visualization?</h2>

			<div class="screenshot"><img src="images/screen1.png" alt="Screenshot 1"/></div>
			<p>Here's the main view, where you can see the genome with connecting arcs between chromosomes.</p>

			<div class="screenshot"><img src="images/screen2.png" alt="Screenshot 2"/></div>
			<p>For a better view chromosomes can be minimized or maximized from the context menu that opens with right click. 'Restore all' restores all chromosomes to their original size.<br/>
			'Fullscreen' from the context menu or pressing 'F' from keyboard changes to fullscreen view.</p>

			<div class="screenshot"><img src="images/screen3.png" alt="Screenshot 3"/></div>
			<p>To select a connection to view, left click a place on the genome wheel, where you want to limit the selection to and yellow selector lines limiting the shown connections appear. Now the wanted connection is easier to select. Select with left mouse click or with keyboard with left and right arrows and enter. Also the size of the limited area can be adjusted with mouse wheel.</p>

			<div class="screenshot"><img src="images/screen4.png" alt="Screenshot 4"/></div>
			<p>The selected places (the ends of the selected arc) from the chromosome open into a trackview, where more data can be seen. Trackview can be zoomed with the mouse wheel. To return to the circle view click the x that closes the window.</p>

			<div class="screenshot"><img src="images/screen5.png" alt="Screenshot 5"/></div>
			<p>The opened connection remains as a minimized capsule that can be left clicked to reopen the view or closed by right clicking. The line from the capsule points on the view's place on the chromosome. New minimized capsules can also be opened by left clicking outside the circle.</p>

			<h3>Command line parameters</h3>

			<pre><code>-f [number]  		link filtering range (0 turns filtering off)
-g [h/r]	      	human or rat genome data
-bam [bam-file]	opens a bam-file
-bai [bai-file]	opens a bai-file
-d       		debug-mode
-p [data path]         sets the data path
</code></pre>
			<nav id="download"><a href="#index">Index</a> - <a href="#video">Usage video</a> - <a href="#howto">How to use</a> - <a href="#download">Download</a></nav>
			<h2>Download</h2>
			<p>Note! This page is just for the internal development purposes and you have to wait for the official release to be able to easily run the program.
			<p>Grab your own copy from <a href="jar">here</a> or clone the source code from <a href="http://www.github.com/ohtubrowser/gbrowserng">GitHub</a>.</p>

			<h3>References</h3>
			Data of example screenshots:

			Edgren et al.: Identification of fusion genes in breast cancer by paired-end RNA-sequencing. Genome Biology 2011 12:R6. <br>
			Sequence Read Archive Accession: SRR064438 paired-end sequencing of BT-474 cDNA

			<!--
			<h3>Authors and Contributors</h3>

			<p>You can <a href="https://github.com/blog/821" class="user-mention">@mention</a> a GitHub username to generate a link to their profile. The resulting <code>&lt;a&gt;</code> element will link to the contributor's GitHub Profile. For example: In 2007, Chris Wanstrath (<a href="https://github.com/defunkt" class="user-mention">@defunkt</a>), PJ Hyett (<a href="https://github.com/pjhyett" class="user-mention">@pjhyett</a>), and Tom Preston-Werner (<a href="https://github.com/mojombo" class="user-mention">@mojombo</a>) founded GitHub.</p>
			-->
			</section>
		</div>
		<!--[if !IE]><script>fixScale(document);</script><!--<![endif]-->

	</body>
</html>
